rfcs/20210731-tfjs-named-tensors.md

```diff
@@ -0,0 +1,557 @@
+  valid dimensions. Referencing an index that does not exist can no longer
+  happen. e.g. auto-completion of `xs.` results in `inputRep` and `seqLen`. No
+  need to remember which order they have either.
+* The optimal ordering of indexes can be computed behind the scenes, so the user
+  does not need to worry about transpositions and permutations, or how to make
+  it optimal.
+
+## User Benefit
+
+A headline might be: "ML programming in JS just got much easier with Named Dimensions". Programming is easier in two import ways:
+
+1. It's easier to write code (and harder to write subtle bugs). This is because
+   there is helpful auto-completion, one doesn't need to remember the
+   dimensions, and you can no longer accidentally multiply the wrong dimensions
+   together (you can do this only via an explicit rename operation).
+2. And It's easier to read code too: one doesn't need to remember and map the
+   indexes to what they mean.
+3. You don't need to work hard to find the optimal ordering and calling of
+   permutation operations. This gets done for you behind the scenes.
+
+This combines many of the benefits of eager and lazy evaluation.
+
+## Design Proposal
+
+The key idea is to introduce a Named-Dimension layer over the TFJS Tensor
+object. This maintains a mapping between the indexes and their corresponding
+names. We leverage TypeScript's ability to have a type that is a set of string
+literals (within union operators `|` ). Using TypeScripts intersection (`&`) and
+exclusion (`Exclude`) type operations, we can automatically infer types
+appropriately. We can also use the type choice operator (`?`) and introduce some
+unique error types to make error messages easy to read and act upon.
+
+A `GTensor` (a "Graph Tensor", inspired by
+[OpenGraphs](https://arxiv.org/abs/1011.4114)) is a tensor with a set of named
+dimensions parameters (each dimension is an axis in the larger tensor). Each
+named dimension parameter is just the literal string type of the name.
+
+Each `GTensor` object has a `dim` property that gives access to the dimensions
+as named properties. Each dimension is parameterized by both the `GTensor` it
+comes from, as well as the name of the type it is a dimension of. To select
+which dimension to perform an operation on, one simply uses the property of that
+dimension.
+
+```ts
+const g1: GTensor<'inputTokens'|'tokenRep'> = ...;
+const g2: GTensor<'tokenRep'|'queryRep'> = ...;
+const g3 = g1.dim.tokenRep.dot(g2.dim.tokenRep);
+// g3: GTensor<'inputTokens'|'queryRep'>
+```
+
+Type-checking ensures that dimension names match. i.e.
+
+```ts
+g1.dim.tokenRep.dot(g2.dim.foo)
+             // Type Error: "tokenRep" is different from "foo".
+```
+
+Dimensions can be renamed also to provide a new dimension object with the
+correct name, e.g. if `g2` didn't have an `tokenRep` dimension, but we wanted to
+dot product with the `foo` dimension, we could do:
+
+```ts
+g1.dim.tokenRep.dot(g2.dim.foo.rename('tokenRep'));
```

Comment:
This seems clunky and confusing.  Maybe you all have thought this through more and I understand what you are trying to do, but this line seems to be adding complexity while other aspects, I agree, are improving things.  I just don't see how this is improving readability or reducing complexity.  I realize that's subjective.  Perhaps a more concrete example (don't use 'foo' as a dimension).